File: src/systems/TrafficAlertAdapter.js
import {
  ALERT_DURATION_KINDS,
  ALERT_FOCUS_ACTIONS,
  ALERT_SEVERITIES,
  ALERT_TYPES
} from '../alerts/AlertService.js';

function bridgeFocus() {
  return {
    type: ALERT_FOCUS_ACTIONS.MANAGEMENT_CAMERA,
    position: { x: 155, y: 0, z: 0 }
  };
}

/** Publishes structured alerts from aggregate mobility state changes. */
export class TrafficAlertAdapter {
  constructor({ model, alertService } = {}) {
    if (!model?.subscribe) throw new TypeError('model must expose subscribe()');
    if (!alertService?.publish || !alertService?.resolve) {
      throw new TypeError('alertService must expose publish() and resolve()');
    }
    this.model = model;
    this.alertService = alertService;
    this.unsubscribe = model.subscribe(event => this.sync(event.current), { emitCurrent: true });
  }

  sync(snapshot) {
    if (!snapshot) return false;
    const congestionKey = 'traffic:network-congestion';
    if (snapshot.network.congestion >= 0.5) {
      const hotspot = snapshot.network.hotspots[0];
      this.alertService.publish({
        dedupeKey: congestionKey,
        type: ALERT_TYPES.TRAFFIC,
        severity: snapshot.network.congestion >= 0.75
          ? ALERT_SEVERITIES.CRITICAL
          : ALERT_SEVERITIES.WARNING,
        title: 'Road congestion is reducing city productivity',
        cause: `${Math.round(snapshot.network.congestion * 100)}% aggregate congestion leaves productivity at ${snapshot.productivity.percent}% and ${snapshot.jobs.jobsDelayedByCommute.toLocaleString()} jobs commute-limited.`,
        location: hotspot
          ? { label: hotspot.label, position: { x: hotspot.x, y: 0, z: hotspot.z } }
          : 'City road network',
        duration: { kind: ALERT_DURATION_KINDS.UNTIL_RESOLVED },
        recommendation: 'Clear incidents, connect road segments, or review bridge priority and its disclosed tradeoff.',
        relatedEntityIds: hotspot ? [hotspot.id] : [],
        focusAction: hotspot
          ? { type: ALERT_FOCUS_ACTIONS.MANAGEMENT_CAMERA, position: { x: hotspot.x, y: 0, z: hotspot.z } }
          : { type: ALERT_FOCUS_ACTIONS.NONE }
      });
    } else {
      this.alertService.resolve(congestionKey, 'Aggregate congestion returned below the disruption threshold');
    }

    const bridgeKey = 'traffic:bridge-disruption';
    if (snapshot.bridge.outageActive || snapshot.bridge.access !== 'OPEN') {
      this.alertService.publish({
        dedupeKey: bridgeKey,
        type: ALERT_TYPES.TRAFFIC,
        severity: snapshot.bridge.access === 'CLOSED'
          ? ALERT_SEVERITIES.CRITICAL
          : ALERT_SEVERITIES.WARNING,
        title: snapshot.bridge.access === 'CLOSED' ? 'Primary bridge closed' : 'Primary bridge flow restricted',
        cause: `${snapshot.bridge.streetStatus}; on-time deliveries are ${snapshot.deliveries.onTimePercent}%.`,
        location: { label: 'Primary bridge corridor', position: { x: 155, y: 0, z: 0 } },
        duration: { kind: ALERT_DURATION_KINDS.UNTIL_RESOLVED },
        recommendation: snapshot.bridge.outageActive
          ? 'Fund and complete the linked street repair work before relying on bridge priority.'
          : 'Inspect the bridge restriction and restore access.',
        relatedEntityIds: ['primary-bridge', ...snapshot.bridge.outageIds],
        focusAction: bridgeFocus()
      });
    } else {
      this.alertService.resolve(bridgeKey, 'Primary bridge access and service returned to normal');
    }

    const deliveryKey = 'traffic:delivery-reliability';
    if (snapshot.deliveries.reliability < 0.72) {
      this.alertService.publish({
        dedupeKey: deliveryKey,
        type: ALERT_TYPES.ECONOMY,
        severity: ALERT_SEVERITIES.WARNING,
        title: 'Delivery backlog is creating contract demand',
        cause: `${snapshot.deliveries.delayedPercent}% of aggregate deliveries are delayed by road and bridge conditions.`,
        location: 'Citywide logistics network',
        duration: { kind: ALERT_DURATION_KINDS.UNTIL_RESOLVED },
        recommendation: 'Freight contracts now pay a disruption premium; freight priority improves reliability but costs $120/min and 2 satisfaction.',
        relatedEntityIds: [],
        focusAction: bridgeFocus()
      });
    } else {
      this.alertService.resolve(deliveryKey, 'Delivery reliability recovered above the backlog threshold');
    }
    return true;
  }

  dispose() {
    this.unsubscribe?.();
  }
}

export default TrafficAlertAdapter;

File: godot/MetroPulse.Domain/Alerts/TrafficAlertAdapter.cs
using System.Globalization;
using MetroPulse.Domain.Traffic;

namespace MetroPulse.Domain.Alerts;

/// <summary>Publishes structured alerts from authoritative aggregate mobility changes.</summary>
public sealed class TrafficAlertAdapter : IDisposable
{
    private readonly AlertService alerts;
    private readonly Func<bool> unsubscribe;

    public TrafficAlertAdapter(TrafficProductivityModel model, AlertService alerts)
    {
        ArgumentNullException.ThrowIfNull(model);
        this.alerts = alerts ?? throw new ArgumentNullException(nameof(alerts));
        unsubscribe = model.Subscribe(trafficEvent => Sync(trafficEvent.Current), emitCurrent: true);
    }

    public bool Sync(TrafficProductivitySnapshot snapshot)
    {
        ArgumentNullException.ThrowIfNull(snapshot);
        SyncCongestion(snapshot);
        SyncBridge(snapshot);
        SyncDeliveries(snapshot);
        return true;
    }

    public void Dispose() => unsubscribe();

    private void SyncCongestion(TrafficProductivitySnapshot snapshot)
    {
        const string key = "traffic:network-congestion";
        if (snapshot.Network.Congestion < 0.5)
        {
            alerts.Resolve(key, "Aggregate congestion returned below the disruption threshold");
            return;
        }

        TrafficHotspot? hotspot = snapshot.Network.Hotspots.FirstOrDefault();
        AlertLocation location = hotspot is null
            ? AlertLocation.Named("City road network")
            : new AlertLocation
            {
                Label = hotspot.Label,
                Position = new AlertPosition(hotspot.X, hotspot.Z),
            };
        alerts.Publish(new AlertInput
        {
            DedupeKey = key,
            Type = AlertTypes.Traffic,
            Severity = snapshot.Network.Congestion >= 0.75 ? AlertSeverities.Critical : AlertSeverities.Warning,
            Title = "Road congestion is reducing city productivity",
            Cause = $"{JavascriptRound(snapshot.Network.Congestion * 100)}% aggregate congestion leaves productivity at {snapshot.Productivity.Percent}% and {snapshot.Jobs.JobsDelayedByCommute.ToString("N0", CultureInfo.InvariantCulture)} jobs commute-limited.",
            Location = location,
            Duration = new AlertDuration { Kind = AlertDurationKinds.UntilResolved },
            Recommendation = "Clear incidents, connect road segments, or review bridge priority and its disclosed tradeoff.",
            RelatedEntityIds = hotspot is null ? Array.Empty<string>() : new[] { hotspot.Id },
            FocusAction = hotspot is null
                ? new AlertFocusAction()
                : new AlertFocusAction
                {
                    Type = AlertFocusActions.ManagementCamera,
                    Position = new AlertPosition(hotspot.X, hotspot.Z),
                },
        });
    }

    private void SyncBridge(TrafficProductivitySnapshot snapshot)
    {
        const string key = "traffic:bridge-disruption";
        if (!snapshot.Bridge.OutageActive && snapshot.Bridge.Access == TrafficAccess.Open)
        {
            alerts.Resolve(key, "Primary bridge access and service returned to normal");
            return;
        }

        alerts.Publish(new AlertInput
        {
            DedupeKey = key,
            Type = AlertTypes.Traffic,
            Severity = snapshot.Bridge.Access == TrafficAccess.Closed ? AlertSeverities.Critical : AlertSeverities.Warning,
            Title = snapshot.Bridge.Access == TrafficAccess.Closed ? "Primary bridge closed" : "Primary bridge flow restricted",
            Cause = $"{snapshot.Bridge.StreetStatus}; on-time deliveries are {snapshot.Deliveries.OnTimePercent}%.",
            Location = BridgeLocation(),
            Duration = new AlertDuration { Kind = AlertDurationKinds.UntilResolved },
            Recommendation = snapshot.Bridge.OutageActive
                ? "Fund and complete the linked street repair work before relying on bridge priority."
                : "Inspect the bridge restriction and restore access.",
            RelatedEntityIds = new[] { TrafficProductivityModel.PrimaryBridgeId }.Concat(snapshot.Bridge.OutageIds).ToArray(),
            FocusAction = BridgeFocus(),
        });
    }

    private void SyncDeliveries(TrafficProductivitySnapshot snapshot)
    {
        const string key = "traffic:delivery-reliability";
        if (snapshot.Deliveries.Reliability >= 0.72)
        {
            alerts.Resolve(key, "Delivery reliability recovered above the backlog threshold");
            return;
        }

        alerts.Publish(new AlertInput
        {
            DedupeKey = key,
            Type = AlertTypes.Economy,
            Severity = AlertSeverities.Warning,
            Title = "Delivery backlog is creating contract demand",
            Cause = $"{snapshot.Deliveries.DelayedPercent}% of aggregate deliveries are delayed by road and bridge conditions.",
            Location = AlertLocation.Named("Citywide logistics network"),
            Duration = new AlertDuration { Kind = AlertDurationKinds.UntilResolved },
            Recommendation = "Freight contracts now pay a disruption premium; freight priority improves reliability but costs $120/min and 2 satisfaction.",
            RelatedEntityIds = Array.Empty<string>(),
            FocusAction = BridgeFocus(),
        });
    }

    private static AlertLocation BridgeLocation() => new()
    {
        Label = "Primary bridge corridor",
        DistrictId = TrafficProductivityModel.PrimaryBridgeDistrict,
        Position = new AlertPosition(155, 0),
    };

    private static AlertFocusAction BridgeFocus() => new()
    {
        Type = AlertFocusActions.ManagementCamera,
        Position = new AlertPosition(155, 0),
    };

    private static int JavascriptRound(double value) => checked((int)Math.Floor(value + 0.5));
}
